packages/cicero-core/src/contractinstance.js

```diff
@@ -0,0 +1,230 @@
+     * @param {String} p12File - encoded string of p12 keystore file
+     * @param {String} passphrase - passphrase for the keystore file
+     * @param {String} signatory - name of the signatory
+     * @return {Promise<Buffer>} the zlib buffer
+     */
+    async signContract(p12File, passphrase, signatory) {
+        if (this.contractSignatures.length !== 0 ) {
+            this.verifySignatures();
+        }
+        const timestamp = Date.now();
+        this.sign(p12File, passphrase, timestamp, signatory);
+
+        //Add state
+        Util.addHistory(
+            this,
+            signatory,
+            'sign',
+            'Signed Succesfully',
+            'Signing in Progress'
+        );
+
+        return await this.toArchive('ergo');
+    }
+
+    /**
+     * signs a string made up of contract hash and time stamp using private key derived
+     * from the keystore
+     * @param {String} p12File - encoded string of p12 keystore file
+     * @param {String} passphrase - passphrase for the keystore file
+     * @param {Number} timestamp - timestamp of the moment of signature is done
+     * @param {String} signatory - name of the signatory
+     * @ppublic
+     */
+    sign(p12File, passphrase, timestamp, signatory) {
+        if (typeof(p12File) !== 'string') {throw new Error('p12File should be of type String!');}
+        if (typeof(passphrase) !== 'string') {throw new Error('passphrase should be of type String!');}
+        if (typeof(timestamp) !== 'number') {throw new Error('timeStamp should be of type Number!');}
+
+        const instanceHash = this.getHash();
+        // decode p12 from base64
+        const p12Der = forge.util.decode64(p12File);
+        // get p12 as ASN.1 object
+        const p12Asn1 = forge.asn1.fromDer(p12Der);
+        // decrypt p12 using the passphrase 'password'
+        const p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, false, passphrase);
+        //X509 cert forge type
+        const certificateForge = p12.safeContents[0].safeBags[0].cert;
+        //Private Key forge type
+        const privateKeyForge = p12.safeContents[1].safeBags[0].key;
+        //convert cert and private key from forge to PEM
+        const certificatePem = forge.pki.certificateToPem(certificateForge);
+        const privateKeyPem = forge.pki.privateKeyToPem(privateKeyForge);
+        //convert private key in pem to private key type in node
+        const privateKey = crypto.createPrivateKey(privateKeyPem);
+        const sign = crypto.createSign('SHA256');
+        sign.write(instanceHash + timestamp);
+        sign.end();
+        const signature = sign.sign(privateKey, 'hex');
+        const signatureObject = {
+            signatory,
+            instanceHash,
+            timestamp,
+            signatoryCert: certificatePem,
+            signature
+        };
+        this.contractSignatures.push(signatureObject);
+    }
+
+    /**
+     * Gets a content based SHA-256 hash for this contract instance. Hash
```

Comment:
This should use the existing hash for the template - which is based on the template metadata, models and ergo files. Lets update that code is necessary rather than redefining the hash here.

Reply:
Maybe move some of this shared code logic to a `cicero-utils` package/class?